packages/block-editor/src/components/block-draggable/index.native.js

```diff
@@ -0,0 +1,315 @@
+				}
+			>
+				{ children( { onScroll: scrollHandler } ) }
+			</Draggable>
+			<Animated.View
+				onLayout={ onChipLayout }
+				style={ chipStyles }
+				pointerEvents="none"
+			>
+				<DraggableChip />
+			</Animated.View>
+		</>
+	);
+};
+
+/**
+ * Block draggable component
+ *
+ * This component serves for animating the block when it is being dragged.
+ * Hence, it should be wrapped around the rendering of a block.
+ *
+ * @param {Object}      props          Component props.
+ * @param {JSX.Element} props.children Children to be rendered.
+ * @param {string[]}    props.clientId Client id of the block.
+ *
+ * @return {Function} Render function which includes the parameter `isDraggable` to determine if the block can be dragged.
+ */
+const BlockDraggable = ( { clientId, children } ) => {
+	const container = {
+		height: useSharedValue( 0 ),
+	};
+	const containerHeightBeforeDragging = useSharedValue( 0 );
+	const collapseAnimation = useSharedValue( 0 );
+
+	const onContainerLayout = ( { nativeEvent: { layout } } ) => {
+		container.height.value = layout.height;
+	};
+
+	const startBlockDragging = () => {
+		'worklet';
+		containerHeightBeforeDragging.value = container.height.value;
+		collapseAnimation.value = withTiming( 1 );
+	};
+
+	const stopBlockDragging = () => {
+		'worklet';
+		collapseAnimation.value = withTiming( 0 );
+	};
+
+	const { isDraggable, isBeingDragged } = useSelect(
+		( select ) => {
+			const {
+				getBlockRootClientId,
+				getTemplateLock,
+				isBlockBeingDragged,
+			} = select( blockEditorStore );
+			const rootClientId = getBlockRootClientId( clientId );
+			const templateLock = rootClientId
+				? getTemplateLock( rootClientId )
+				: null;
+
+			return {
+				isBeingDragged: isBlockBeingDragged( clientId ),
+				isDraggable: 'all' !== templateLock,
+			};
+		},
+		[ clientId ]
+	);
+
+	useEffect( () => {
+		if ( isBeingDragged ) {
+			runOnUI( startBlockDragging )();
+		} else {
+			runOnUI( stopBlockDragging )();
+		}
+	}, [ isBeingDragged ] );
+
+	const containerStyles = useAnimatedStyle( () => {
+		const height = interpolate(
+			collapseAnimation.value,
+			[ 0, 1 ],
+			[ containerHeightBeforeDragging.value, BLOCK_COLLAPSED_HEIGHT ]
+		);
+		return {
+			height:
+				containerHeightBeforeDragging.value === 0 ||
+				collapseAnimation.value === 0
+					? 'auto'
+					: height,
+		};
+	} );
+
+	const blockStyles = useAnimatedStyle( () => {
+		return {
+			opacity: 1 - collapseAnimation.value,
+		};
+	} );
+
+	const placeholderStyles = useAnimatedStyle( () => {
+		return {
+			display: collapseAnimation.value === 0 ? 'none' : 'flex',
+			opacity: collapseAnimation.value,
+		};
+	} );
+
+	if ( ! isDraggable ) {
+		return children( { isDraggable: false } );
+	}
+
+	return (
+		<Animated.View onLayout={ onContainerLayout } style={ containerStyles }>
+			<Animated.View style={ blockStyles }>
+				{ children( { isDraggable: true } ) }
+			</Animated.View>
+			<Animated.View
+				style={ [
```

Comment:
Would it make sense to have the styles composition in a variable outside of the render?

Reply:
I applied this suggestion in https://github.com/WordPress/gutenberg/pull/39617/commits/4060e1ad86e6d2c9b880a7ed28509668c146f062.